sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -0,0 +1,81 @@
+package compliance
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/concurrency"
+)
+
+type nodeScanHandlerImpl struct {
+	nodeScans <-chan *storage.NodeScanV2
+	toCentral chan *central.MsgFromSensor
+
+	stopC    concurrency.ErrorSignal
+	stoppedC concurrency.ErrorSignal
+}
+
+func (c *nodeScanHandlerImpl) Capabilities() []centralsensor.SensorCapability {
+	return []centralsensor.SensorCapability{centralsensor.NodeScanningV2}
+}
+
+func (c *nodeScanHandlerImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	return c.toCentral
+}
+
+func (c *nodeScanHandlerImpl) Start() error {
+	go c.run()
+	return nil
+}
+
+func (c *nodeScanHandlerImpl) Stop(err error) {
+	c.stopC.SignalWithError(err)
+}
+
+func (c *nodeScanHandlerImpl) Stopped() concurrency.ReadOnlyErrorSignal {
+	return &c.stoppedC
+}
+
+func (c *nodeScanHandlerImpl) ProcessMessage(msg *central.MsgToSensor) error {
+	// This component doesn't actually process or handle any messages sent from Central to Sensor (yet).
+	// It uses the sensor component so that the lifecycle (start, stop) can be handled when Sensor starts up.
+	return nil
+}
+
+func (c *nodeScanHandlerImpl) run() {
+	defer c.stoppedC.Signal()
+	for {
+		select {
+		case <-c.stopC.Done():
+			c.stoppedC.SignalWithError(c.stopC.Err())
+
+		case scan, ok := <-c.nodeScans:
+			if !ok {
+				c.stoppedC.SignalWithError(errors.New("channel receiving node scans v2 is closed"))
+				return
+			}
+			// Do something with the scan if needed, e.g. attach NodeID
+			c.sendScan(scan)
+		}
+	}
+}
+
+func (c *nodeScanHandlerImpl) sendScan(scan *storage.NodeScanV2) {
+	select {
+	case <-c.stoppedC.Done():
```

Comment:
The idea of this case is not very clear to me. When the stop intent is registered, we eat up messages from `c.nodeScans` but drop them on the floor instead of sending to `c.toCentral`. Is that your idea? Would it make sense to just not read messages from `c.nodeScans` at all instead of dropping them?

Reply:
This was a copy-paste, but I think that we do not send the message because it may happen that the stream will be  already closed when the context is canceled. 
I think the risk here is that the `c.toCentral` channel may not have any readers attached when `c.stoppedC.Done()` happens. This could lead to the `c.toCentral <- &central.MsgFromSensor{` blocking despite the context has been already canceled.

I think this deserves an experiment to be sure, WDYT?

Reply:
I guess my question is more about what should we expect from `c.nodeScans`. Depending on that we can make decisions what to do with `c.toCentral`.

Also, I wasn't sure if we should rather handle `c.stopC.Done()` instead of `c.stoppedC.Done()` here.

Reply:
I will look into `stopC` and `stoppedC` (including your [other comment](https://github.com/stackrox/stackrox/pull/3533#discussion_r1011511017)).

`c.nodeScans` is created in the compliance service `compliance/service_impl.go` and it receives all messages of type `*sensor.MsgFromCompliance_NodeScanV2` directly from the stream (`msg, err := server.Recv()`).
Unfortunately, `c.nodeScans` is never closed, as we do not have a function that would be called when the service is shutdown. We do not want to close the channel when the stream is closed, because the stream can be reopened. We should close the channel on sensor shutdown, but the API of `type Service interface` does not have any method that can be called to signal this.

Reply:
I played a bit with stopping in the test that I added in [ab64a00](https://github.com/stackrox/stackrox/pull/3533/commits/ab64a0004f037dbfaeade9b899dc58b3270b4781). This allowed me to simplify the `sendScan`, so that we are not dropping anything. 

I will give it a test in a local cluster, but the unit test that I added looks good!

Reply:
> I will give it a test in a local cluster, but the unit test that I added looks good!

I have discovered no issues.